### PR TITLE
Adjust interception of requests for specific shard IDs

### PR DESCRIPTION
**Pull request by AthenaEryma**
Some index requests target shard IDs specifically, which may not match the indices that the request targets as given by IndicesRequest, which requires a different interception strategy in order to make sure those requests are handled correctly in all cases.

This PR adds and interface allowing requests to report the shard IDs they target as well as the index names, and adjusts the interception of those requests as appropriate to handle those shard IDs in the cases where they are relevant.

### CODE REVIEW

**Review by kpollich — APPROVED** [review]
Fleet codeowner 🚀

**Comment by gwbrown** [comment]
@jakelandis I believe I've addressed the concerns we discussed, could you give me a review?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/support/CrossClusterShardTests.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        RetentionLeaseActions.REMOVE.name(),
+        RetentionLeaseActions.RENEW.name(),
+        TermVectorsAction.NAME,
+        TransportGetAction.TYPE.name(),
+        TransportShardMultiGetAction.TYPE.name(),
+        TransportGetFieldMappingsIndexAction.TYPE.name(),
+
+        // These actions do not have any references to shard IDs in their requests.
+        ClusterSearchShardsAction.NAME,
+        TransportSearchShardsAction.NAME
```
Comment (jakelandis) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/support/CrossClusterShardTests.java:55]: this action is associated with ShardSearchRequest which does have shardIDs in their request.

File: test/framework/src/main/java/org/elasticsearch/action/support/single/shard/SingleShardRequestHelper.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+package org.elasticsearch.action.support.single.shard;
+
+import org.elasticsearch.index.shard.ShardId;
+
+public class SingleShardRequestHelper {
+
+    private SingleShardRequestHelper() {}
+
+    public static void setInternalShardId(SingleShardRequest<?> singleShardRequest, ShardId shardId) {
```
Comment (jakelandis) [inline test/framework/src/main/java/org/elasticsearch/action/support/single/shard/SingleShardRequestHelper.java:17]: Seems a bit off to introduce production code to allow invalid states only used in tests. I have an alternative idea will ping you about.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/support/CrossClusterShardTests.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    );
+
+    @Override
+    protected Collection<Class<? extends Plugin>> getPlugins() {
+        return Set.of(LocalStateCompositeXPackPlugin.class);
```
Comment (jakelandis) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/support/CrossClusterShardTests.java:60]: This results in only checking the (core) xpack plugin and (core) server code's registered actions. Do you think it is worth pull this out to it's own qa project and have that qa project depend on a large % of the core/modules and x-pack/plugins. I understand this is best effort only but pulling this out to it's own project will get coverage similar to the default distribution and help to identify any new actions (unless the new actions are part of a new plugin/module)

Comment (gwbrown) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/support/CrossClusterShardTests.java:60]: Blah, whoops, for some reason I thought switching this to `LocalStateCompositeXPackPlugin` addressed that but upon looking again you are correct and it does not. I'll look into this and look to get this pulled out somewhere we get more coverage.

(I really dislike this part of our test infrastructure and find it quite confusing, which plugins are included/accessible where)

**Review by jakelandis — COMMENTED** [review]
Looking good... I wish there were better ways to test, but what you have here is seems the best we can do with what we have. 

A couple comments.

**Comment by gwbrown** [comment]
@jakelandis I yes-and'd idea into "create a writeable that streams the equivalent to a get request"; I think that makes it a bit more maintainable while still letting us remove the test-only production code. I've also expanded the test coverage to cover x-pack, which had a couple implications I Slack'd you about.

**Review by mark-vieira — APPROVED** [review]

**Review by jakelandis — APPROVED** [review]
LGTM

**Pull request merged**